src/compiler/rust/canister_methods/src/hash_file.rs

```diff
@@ -0,0 +1,131 @@
+                        ic_cdk_timers::set_timer(delay, closure);
+                    } else {
+                        // No more bytes to hash, set as final hash for this file
+                        match previous_hash {
+                            Some(hash) => {
+                                set_file_hash(&path, hash);
+                                ic_cdk::println!("Finish hashing");
+                                clear_partial_hash(&path);
+                                ic_cdk::println!("Partial hash cleared");
+                            }
+                            None => ic_cdk::println!("WARNING: No hash was found for {}", path),
+                        }
+                    }
+                }
+                Err(err) => panic!("Error reading file: {}", err),
+            }
+        }
+
+        pub fn get_partial_file_hash(path: &String) -> Option<Vec<u8>> {
+            PARTIAL_FILE_HASHES.with(|file_hashes| file_hashes.borrow().get(path).cloned())
+        }
+
+        fn set_partial_hash(path: &String, hash: Vec<u8>) {
+            PARTIAL_FILE_HASHES.with(|file_hashes| file_hashes.borrow_mut().insert(path.clone(), hash));
+        }
+
+        fn clear_partial_hash(path: &String) {
+            PARTIAL_FILE_HASHES.with(|file_hashes| file_hashes.borrow_mut().remove(path));
+        }
+
+        fn clear_file_hash(path: &String) {
+            let mut file_hashes = load_hashes().unwrap();
+            file_hashes.remove(path);
+            save_hashes(&file_hashes).unwrap();
+        }
+
+        fn set_file_hash(path: &String, hash: Vec<u8>) {
+            let mut file_hashes = load_hashes().unwrap();
+            file_hashes.insert(path.clone(), hash);
+            save_hashes(&file_hashes).unwrap();
+        }
+
+        fn hash_chunk_with(data: &[u8], previous_hash: Option<&Vec<u8>>) -> Vec<u8> {
+            let mut h = <sha2::Sha256 as sha2::Digest>::new();
+            sha2::Digest::update(&mut h, data);
+            if let Some(hash) = previous_hash {
+                sha2::Digest::update(&mut h, hash);
+            }
+            sha2::Digest::finalize(h).to_vec()
+        }
+
+        fn load_hashes() -> Result<HashMap<String, Vec<u8>>, std::io::Error> {
+            let mut file_hashes = HashMap::new();
+            // Check if the file exists
+            if !std::path::Path::new(FILE_HASH_PATH).exists() {
+                return Ok(file_hashes);
+            }
+            let mut file = std::fs::File::open(FILE_HASH_PATH)?;
+
+            let mut data = Vec::new();
+            std::io::Read::read_to_end(&mut file, &mut data)?;
+
+            if !data.is_empty() {
+                file_hashes = serde_json::from_slice(&data)?; // (Example using serde)
+            }
+
+            Ok(file_hashes)
+        }
+
+        fn save_hashes(file_hashes: &HashMap<String, Vec<u8>>) -> Result<(), std::io::Error> {
```

Comment:
Look into simpler ways to read and write

Reply:
apparently std::fs provides some convenience functions that combine std::io::open with std::io::write or std::io::read. I'll use those instead where I can